pkg/apis/pipeline/v1beta1/resultref.go

```diff
@@ -122,12 +133,32 @@ func stripVarSubExpression(expression string) string {
 	return strings.TrimSuffix(strings.TrimPrefix(expression, "$("), ")")
 }
 
-func parseExpression(substitutionExpression string) (string, string, error) {
+// parseExpression parses "task name", "result name", and "attribute name" (iff it's an object result) from the result parseExpression
+// Valid Example 1:
+// - Input: tasks.myTask.results.aStringResult
+// - Output: "myTask", "aStringResult", "", nil
+// Valid Example 2:
+// - Input: tasks.myTask.results.anObjectResult.key1
+// - Output: "myTask", "anObjectResult", "key1", nil
+// Invalid Example 1:
+// - Input: tasks.myTask.results.resultName.foo.bar
+// - Output: "", "", "", error
+func parseExpression(substitutionExpression string) (string, string, string, error) {
 	subExpressions := strings.Split(substitutionExpression, ".")
-	if len(subExpressions) != 4 || subExpressions[0] != ResultTaskPart || subExpressions[2] != ResultResultPart {
-		return "", "", fmt.Errorf("Must be of the form %q", resultExpressionFormat)
+
+	// TODO (@yongxuanzhang https://github.com/tektoncd/pipeline/issues/4723):
+	// consider array type result here tasks.<taskName>.results.<resultName>[*] and indexing tasks.<taskName>.results.<resultName>[0]
+	// For string type result: tasks.<taskName>.results.<resultName>
+	if len(subExpressions) == 4 && subExpressions[0] == ResultTaskPart && subExpressions[2] == ResultResultPart {
+		return subExpressions[1], subExpressions[3], "", nil
 	}
-	return subExpressions[1], subExpressions[3], nil
+
+	// For object type result: tasks.<taskName>.results.<objectResultName>.<individualAttribute>
+	if len(subExpressions) == 5 && subExpressions[0] == ResultTaskPart && subExpressions[2] == ResultResultPart {
+		return subExpressions[1], subExpressions[3], subExpressions[4], nil
+	}
+
+	return "", "", "", fmt.Errorf("Must be one of the form 1). %q; 2). %q", resultExpressionFormat, objectResultExpressionFormat)
```

Comment:
Just added a test case for too many dot-separated components.
The test cases with other invalid format already exist in the function named `TestNewResultReference` which tests its caller, but its caller is expected to return empty resultref rather than an error. So that's why this message will not be printed out.